cpp/src/gandiva/expression_registry.h

```diff
@@ -62,7 +64,7 @@ class GANDIVA_EXPORT ExpressionRegistry {
 
  private:
   static DataTypeVector supported_types_;
-  std::unique_ptr<FunctionRegistry> function_registry_;
```

Comment:
Apropos ownership, do we want to carry raw pointers? Using `std::shared_ptr` may be safer and more future-proof...

Reply:
I don't change this part so far, currently, `gandiva::default_function_registry()` returns a `FunctionRegistry *` raw pointer, and the `ExpressionRegistry` uses `gandiva::default_function_registry()` as the default value for its internal `function_registry_` variable. If we use `std::shared_ptr` here, do you have any idea what we should use as the default value for `function_registry_` variable in `ExpressionRegistry`? Thanks.

Reply:
Edit: `default_function_registry` is a new API, so we can freely make it return a `std::shared_ptr`.

Reply:
>  If default_function_registry is already a public API

No, it is not a public API yet (it is introduced in this PR), so I think we still have the chance to change its signature if needed. Currently, this `gandiva::default_function_registry()` function's signature and usage mimics `arrow::default_memory_pool()` which returns a `MemoryPool *` raw pointer. Do you mean we should change this API to return a `std::shared_ptr<FunctionRegistry>` like this:
```C++
namespace gandiva {
  std::shared_ptr<FunctionRegistry> default_function_registry();
}
```

Reply:
Okay, I will update the `default_function_registry` API to return a `std::shared_ptr` instead.

Reply:
I updated the `default_function_registry` and several usages like `ExpressionRegistry` to use `std::shared_ptr`.